cpp/oneapi/dal/backend/primitives/rng/dpc_engine.hpp

```diff
@@ -0,0 +1,233 @@
+/// some engine types.
+template <engine_method EngineType = engine_method::mt2203>
+class dpc_engine {
+public:
+    using dpc_engine_t = typename dpc_engine_type<EngineType>::type;
+
+    explicit dpc_engine(sycl::queue& queue, std::int64_t seed = 777)
+            : q(queue),
+              host_engine_(initialize_host_engine(seed)),
+              dpc_engine_(initialize_dpc_engine(queue, seed)),
+              impl_(dynamic_cast<daal::algorithms::engines::internal::BatchBaseImpl*>(
+                  host_engine_.get())) {
+        if (!impl_) {
+            throw std::domain_error("RNG engine is not supported");
+        }
+    }
+
+    virtual ~dpc_engine() = default;
+
+    void* get_host_engine_state() const {
+        return impl_->getState();
+    }
+
+    auto& get_cpu_engine() {
+        return host_engine_;
+    }
+
+    auto& get_gpu_engine() {
+        return dpc_engine_;
+    }
+
+    void skip_ahead_cpu(size_t nSkip) {
+        host_engine_->skipAhead(nSkip);
+    }
+
+    void skip_ahead_gpu(size_t nSkip) {
+        // Will be supported in the next oneMKL release.
+        if constexpr (EngineType == engine_method::mt2203) {
+        }
+        else {
+            skip_ahead(dpc_engine_, nSkip);
+        }
+    }
+
+    sycl::queue& get_queue() {
+        return q;
+    }
+
+private:
+    daal::algorithms::engines::EnginePtr initialize_host_engine(std::int64_t seed) {
+        switch (EngineType) {
+            case engine_method::mt2203:
+                return daal::algorithms::engines::mt2203::Batch<>::create(seed);
+            case engine_method::mcg59:
+                return daal::algorithms::engines::mcg59::Batch<>::create(seed);
+            case engine_method::mrg32k3a:
+                return daal::algorithms::engines::mrg32k3a::Batch<>::create(seed);
+            case engine_method::philox4x32x10:
+                return daal::algorithms::engines::philox4x32x10::Batch<>::create(seed);
+            case engine_method::mt19937:
+                return daal::algorithms::engines::mt19937::Batch<>::create(seed);
+            default: throw std::invalid_argument("Unsupported engine type");
+        }
+    }
+
+    dpc_engine_t initialize_dpc_engine(sycl::queue& queue, std::int64_t seed) {
+        if constexpr (EngineType == engine_method::mt2203) {
+            return dpc_engine_t(
+                queue,
+                seed,
+                0); // Aligns CPU and GPU results for mt2203, impacts the performance.
+        }
+        else {
+            return dpc_engine_t(queue, seed);
+        }
+    }
+    sycl::queue q;
+    daal::algorithms::engines::EnginePtr host_engine_;
+    dpc_engine_t dpc_engine_;
+    daal::algorithms::engines::internal::BatchBaseImpl* impl_;
+};
+
+template <typename Type, typename Size, engine_method EngineType>
+void uniform(Size count, Type* dst, dpc_engine<EngineType>& engine_, Type a, Type b) {
```

Comment:
Is it necessary to have `Size` as a template parameter?

Reply:
Please provide descriptions to all those APIs. Because from the first glance it is not clear which of them are for CPU only.